### ISSUE TITLE
PHP Deprecated:  Return type of Requests_* on 8.1 with PHPStan

### ISSUE THREAD

**Issue by Mte90** [php-stubs/wordpress-stubs#41]
```
PHP Deprecated:  Return type of Requests_Utility_FilteredIterator::unserialize($serialized) should either be compatible with ArrayIterator::unserialize(string $data): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /home/www/VVV/www/woocommerce/public_html/wp-content/plugins/plugin/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 21682
Deprecated: Return type of Requests_Utility_FilteredIterator::unserialize($serialized) should either be compatible with ArrayIterator::unserialize(string $data): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /home/www/VVV/www/woocommerce/public_html/wp-content/plugins/plugin/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 21682
PHP Deprecated:  Return type of Requests_Utility_FilteredIterator::__unserialize($serialized) should either be compatible with ArrayIterator::__unserialize(array $data): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /home/www/VVV/www/woocommerce/public_html/wp-content/plugins/plugin/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 21690
Deprecated: Return type of Requests_Utility_FilteredIterator::__unserialize($serialized) should either be compatible with ArrayIterator::__unserialize(array $data): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /home/www/VVV/www/woocommerce/public_html/wp-content/plugins/plugin/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 21690
PHP Deprecated:  Return type of Requests_Utility_FilteredIterator::current() should either be compatible with ArrayIterator::current(): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /home/www/VVV/www/woocommerce/public_html/wp-content/plugins/plugin/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 21676
Deprecated: Return type of Requests_Utility_FilteredIterator::current() should either be compatible with ArrayIterator::current(): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /home/www/VVV/www/woocommerce/public_html/wp-content/plugins/plugin/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 21676
PHP Deprecated:  Return type of Requests_Cookie_Jar::offsetExists($key) should either be compatible with ArrayAccess::offsetExists(mixed $offset): bool, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /home/www/VVV/www/woocommerce/public_html/wp-content/plugins/plugin/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 18918
Deprecated: Return type of Requests_Cookie_Jar::offsetExists($key) should either be compatible with ArrayAccess::offsetExists(mixed $offset): bool, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /home/www/VVV/www/woocommerce/public_html/wp-content/plugins/plugin/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 18918
PHP Deprecated:  Return type of Requests_Cookie_Jar::offsetGet($key) should either be compatible with ArrayAccess::offsetGet(mixed $offset): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /home/www/VVV/www/woocommerce/public_html/wp-content/plugins/plugin/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 18927
Deprecated: Return type of Requests_Cookie_Jar::offsetGet($key) should either be compatible with ArrayAccess::offsetGet(mixed $offset): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /home/www/VVV/www/woocommerce/public_html/wp-content/plugins/plugin/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 18927
PHP Deprecated:  Return type of Requests_Cookie_Jar::offsetSet($key, $value) should either be compatible with ArrayAccess::offsetSet(mixed $offset, mixed $value): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /home/www/VVV/www/woocommerce/public_html/wp-content/plugins/plugin/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 18938
Deprecated: Return type of Requests_Cookie_Jar::offsetSet($key, $value) should either be compatible with ArrayAccess::offsetSet(mixed $offset, mixed $value): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /home/www/VVV/www/woocommerce/public_html/wp-content/plugins/plugin/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 18938
PHP Deprecated:  Return type of Requests_Cookie_Jar::offsetUnset($key) should either be compatible with ArrayAccess::offsetUnset(mixed $offset): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /home/www/VVV/www/woocommerce/public_html/wp-content/plugins/plugin/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 18946
Deprecated: Return type of Requests_Cookie_Jar::offsetUnset($key) should either be compatible with ArrayAccess::offsetUnset(mixed $offset): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /home/www/VVV/www/woocommerce/public_html/wp-content/plugins/plugin/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 18946
PHP Deprecated:  Return type of Requests_Cookie_Jar::getIterator() should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /home/www/VVV/www/woocommerce/public_html/wp-content/plugins/plugin/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 18954
Deprecated: Return type of Requests_Cookie_Jar::getIterator() should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /home/www/VVV/www/woocommerce/public_html/wp-content/plugins/plugin/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 18954
PHP Deprecated:  Return type of Requests_Utility_CaseInsensitiveDictionary::offsetExists($key) should either be compatible with ArrayAccess::offsetExists(mixed $offset): bool, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /home/www/VVV/www/woocommerce/public_html/wp-content/plugins/plugin/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 20909
Deprecated: Return type of Requests_Utility_CaseInsensitiveDictionary::offsetExists($key) should either be compatible with ArrayAccess::offsetExists(mixed $offset): bool, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /home/www/VVV/www/woocommerce/public_html/wp-content/plugins/plugin/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 20909
PHP Deprecated:  Return type of Requests_Utility_CaseInsensitiveDictionary::offsetGet($key) should either be compatible with ArrayAccess::offsetGet(mixed $offset): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /home/www/VVV/www/woocommerce/public_html/wp-content/plugins/plugin/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 20918
Deprecated: Return type of Requests_Utility_CaseInsensitiveDictionary::offsetGet($key) should either be compatible with ArrayAccess::offsetGet(mixed $offset): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /home/www/VVV/www/woocommerce/public_html/wp-content/plugins/plugin/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 20918
PHP Deprecated:  Return type of Requests_Utility_CaseInsensitiveDictionary::offsetSet($key, $value) should either be compatible with ArrayAccess::offsetSet(mixed $offset, mixed $value): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /home/www/VVV/www/woocommerce/public_html/wp-content/plugins/plugin/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 20929
Deprecated: Return type of Requests_Utility_CaseInsensitiveDictionary::offsetSet($key, $value) should either be compatible with ArrayAccess::offsetSet(mixed $offset, mixed $value): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /home/www/VVV/www/woocommerce/public_html/wp-content/plugins/plugin/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 20929
PHP Deprecated:  Return type of Requests_Utility_CaseInsensitiveDictionary::offsetUnset($key) should either be compatible with ArrayAccess::offsetUnset(mixed $offset): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /home/www/VVV/www/woocommerce/public_html/wp-content/plugins/plugin/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 20937
Deprecated: Return type of Requests_Utility_CaseInsensitiveDictionary::offsetUnset($key) should either be compatible with ArrayAccess::offsetUnset(mixed $offset): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /home/www/VVV/www/woocommerce/public_html/wp-content/plugins/plugin/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 20937
PHP Deprecated:  Return type of Requests_Utility_CaseInsensitiveDictionary::getIterator() should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /home/www/VVV/www/woocommerce/public_html/wp-content/plugins/plugin/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 20945
Deprecated: Return type of Requests_Utility_CaseInsensitiveDictionary::getIterator() should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /home/www/VVV/www/woocommerce/public_html/wp-content/plugins/plugin/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 20945
```

With PHP 8.1 I am getting all those warning. I know that WP itself doesn't support that version but they are very annoying so I wasn't sure if it was to fix there instead on the phpstan-wordpress package.

Maybe it is possible to add `#[\ReturnTypeWillChange]` comment on the stubs for those methods so this error is silenced until WP support that version.

**Comment by szepeviktor** [comment]
> With PHP 8.1

Hello Daniele!
Thank you for your report.

Here are some statements for you.

1. WordPress core is not fully PHP 8.1 compatible
2. The current WordPress ecosystem is not PHP 8.0 compatible
3. WordPress installations live happily on PHP 7.4 as of 2022

> Maybe it is possible to add

This is an automatically generated repo.
If you have a small script or something to fix it please tell me now!

**Comment by Mte90** [comment]
```
#!/usr/bin/env bash

function append_flag() {
    lines=$(awk "/$1/{print FNR}" ./wordpress-stubs.php)
    i=0
    for line in $lines
    do
        echo "Found line $line"
        line=$((line+i))
        if ! [ -z "$2" ]; then
            line=$((line+$2))
        fi
        sed -i $line"i         #[\\ReturnTypeWillChange]" ./wordpress-stubs.php
        echo "Appended to line $line"
        # The line number change after the first insertion
        i=$((i+1))
    done
}

append_flag 'public function unserialize\(\$serialized\)'
append_flag 'public function __unserialize\(\$serialized\)'
append_flag "* Get the current item's value after filtering" 4
append_flag 'public function offsetExists\(\$key\)'
append_flag 'public function offsetGet\(\$key\)'
append_flag 'public function offsetSet\(\$key, \$value\)'
append_flag 'public function offsetUnset\(\$key\)'
append_flag 'public function getIterator\(\)'
```

This bash script fix the stubs and add the flag on the methods reported and now everything is clean :-)

![immagine](https://user-images.githubusercontent.com/403283/179964022-d0de82b4-a8cd-4a5c-bd48-62fff25354aa.png)

The only issue is that is no indentation also if I tried for that but I don't think that is a problem.

Do you want a PR? How I can name the script? `compatibility.sh`?

**Comment by szepeviktor** [comment]
Thank you!

> `for line in $lines`

Do you mean these function appear more than once?

**Comment by Mte90** [comment]
Yes there are methods with the same name that have the same issues in the Request library, this way we fix it with a single line

**Comment by szepeviktor** [comment]
We should create a file similar to https://github.com/php-stubs/wordpress-stubs/blob/master/functionMap.php
and add code to https://github.com/php-stubs/wordpress-stubs/blob/e04781a84e364615a7b5f70fdc345c8253ca5b8f/visitor.php#L37

**Comment by Mte90** [comment]
I don't have time now to convert it in PHP code.

I see that the https://github.com/php-stubs/wordpress-stubs/blob/master/generate.sh#L33 file already do some manipulation by bash so maybe we can add that code there.

**Comment by benjaminprojas** [comment]
I'm still seeing this issue in the latest version. I see the PR was closed... Is there a reason the issue is closed?

**Comment by benjaminprojas** [comment]
The individual notices I am still seeing: 

```
PHP Deprecated:  Return type of Requests_Utility_FilteredIterator::unserialize($serialized) should either be compatible with ArrayIterator::unserialize(string $data): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 21682

PHP Deprecated:  Return type of Requests_Utility_FilteredIterator::__unserialize($serialized) should either be compatible with ArrayIterator::__unserialize(array $data): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 21690

PHP Deprecated:  Return type of Requests_Utility_FilteredIterator::current() should either be compatible with ArrayIterator::current(): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 21676

PHP Deprecated:  Return type of Requests_Cookie_Jar::offsetExists($key) should either be compatible with ArrayAccess::offsetExists(mixed $offset): bool, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 18918

PHP Deprecated:  Return type of Requests_Cookie_Jar::offsetGet($key) should either be compatible with ArrayAccess::offsetGet(mixed $offset): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 18927

PHP Deprecated:  Return type of Requests_Cookie_Jar::offsetSet($key, $value) should either be compatible with ArrayAccess::offsetSet(mixed $offset, mixed $value): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 18938

PHP Deprecated:  Return type of Requests_Cookie_Jar::offsetUnset($key) should either be compatible with ArrayAccess::offsetUnset(mixed $offset): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 18946

PHP Deprecated:  Return type of Requests_Cookie_Jar::getIterator() should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 18954

PHP Deprecated:  Return type of Requests_Utility_CaseInsensitiveDictionary::offsetExists($key) should either be compatible with ArrayAccess::offsetExists(mixed $offset): bool, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 20909

PHP Deprecated:  Return type of Requests_Utility_CaseInsensitiveDictionary::offsetGet($key) should either be compatible with ArrayAccess::offsetGet(mixed $offset): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 20918

PHP Deprecated:  Return type of Requests_Utility_CaseInsensitiveDictionary::offsetSet($key, $value) should either be compatible with ArrayAccess::offsetSet(mixed $offset, mixed $value): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 20929

PHP Deprecated:  Return type of Requests_Utility_CaseInsensitiveDictionary::offsetUnset($key) should either be compatible with ArrayAccess::offsetUnset(mixed $offset): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 20937

PHP Deprecated:  Return type of Requests_Utility_CaseInsensitiveDictionary::getIterator() should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /vendor/php-stubs/wordpress-stubs/wordpress-stubs.php on line 20945

PHP Deprecated:  Optional parameter $id declared before required parameter $object is implicitly treated as a required parameter in /vendor/php-stubs/woocommerce-stubs/woocommerce-stubs.php on line 72705
```

**Comment by janw-me** [comment]
Fixed in #53

**Comment by benjaminprojas** [comment]
Manually including the latest version of the `wordpress-stubs` fixed most of the errors (thanks @janw-me!). The only one remaining is coming from `woocommerce-stubs`.
